source/bolts/experimental/refraction.d

```diff
@@ -0,0 +1,723 @@
+    body_: ";",
+    };
+
+    return model;
+ }
+
+///
+unittest {
+    pure @nogc int answer(lazy string question);
+    alias F = answer; // typically F is a template argument
+    static assert(
+        refract!(F, "F").mixture ==
+        "pure @nogc @system bolts.experimental.refraction.ReturnType!(F) answer(lazy bolts.experimental.refraction.Parameters!(F)[0] _0);");
+}
+
+///
+unittest {
+    import std.format;
+    import std.traits : FunctionAttribute;
+    import bolts.experimental.refraction;
+
+    interface GrandTour {
+        pure int foo() immutable;
+        @nogc @trusted nothrow ref int foo(out real, return ref int, lazy int) const;
+        @safe shared scope void bar(scope Object);
+    }
+
+    class Mock(Interface) : Interface {
+        static foreach (member; __traits(allMembers, Interface)) {
+            static foreach (fun; __traits(getOverloads, Interface, member)) {
+                mixin({
+                        enum Model = refract!(fun, "fun");
+                        if (is(ReturnType!fun == void)) {
+                            return Model.withBody("{}").mixture;
+                        } else if (Model.attributes & FunctionAttribute.ref_) {
+                            return Model.withBody(q{{
+                                        static %s rv;
+                                        return rv;
+                                    }}.format(Model.returnType)).mixture;
+                        } else {
+                            return Model.withBody(q{{
+                                        return %s.init;
+                                    }}.format(Model.returnType)).mixture;
+                        }
+                    }());
+            }
+        }
+    }
+
+    GrandTour mock = new Mock!GrandTour;
+    real x;
+    int i, l;
+    mock.foo(x, i, l++) = 1;
+    assert(mock.foo(x, i, l++) == 1);
+    assert(l == 0);
+}
+
+private enum isFunctionContainer(alias T) = is(T == module) || is(T == struct)
+    || is(T == class) || is(T == interface) || is(T == union);
+
+/**
+   Return a `Function` object refracting the function overload identified by
+   `name` and `index` in `Scope`, using the value of `localName` to represent
+   `Scope`. Store `index` in the `index` property of the Function.
+
+   Params:
+   Scope = a struct, class, interface, or union
+   localName = a string mixin that represents `Scope`
+   name = the name of the function
+   index = the index of the function in the set of overloaded function called `name`
+*/
+
+Function refract(alias Scope, string localName, string name, uint index)()
+if (isFunctionContainer!Scope) {
+    return refract!(
+        __traits(getOverloads, Scope, name)[index],
+        `__traits(getOverloads, %s, "%s")[%d]`.format(localName, name, index))
+        .withIndex(index);
+}
+
+///
+unittest {
+    interface Answers {
+        int answer();
+        string answer();
+    }
+
+    alias Container = Answers;
```

Comment:
Does this have to be aliased or?

Reply:
> Does this have to be aliased or?

It is to emphasize the fact that, usually, the declared name of the container will *not* be used for `localName`.